src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -217,58 +221,703 @@ static void rd_kafka_oidc_build_post_fields(const char *scope,
+        double exp                     = 0;
+        char **extensions              = NULL;
+        char **extension_key_value     = NULL;
+        size_t extension_key_value_cnt = 0;
+        size_t extension_cnt;
+        char *sub = NULL;
+        char validate_errstr[512];
+
+        if (rd_kafka_terminating(rk))
+                return;
+
+        if (rk->rk_conf.sasl.oauthbearer.assertion.file) {
+                jwt_assertion = rd_kafka_oidc_assertion_read_from_file(
+                    rk->rk_conf.sasl.oauthbearer.assertion.file);
+        } else {
+                jwt_assertion = rd_kafka_oidc_assertion_create(
+                    rk, rk->rk_conf.sasl.oauthbearer.assertion.private_key.pem,
+                    rk->rk_conf.sasl.oauthbearer.assertion.private_key.file,
+                    rk->rk_conf.sasl.oauthbearer.assertion.private_key
+                        .passphrase,
+                    rk->rk_conf.sasl.oauthbearer.assertion.algorithm,
+                    rk->rk_conf.sasl.oauthbearer.assertion.jwt_template_file,
+                    rk->rk_conf.sasl.oauthbearer.assertion.claim.subject,
+                    rk->rk_conf.sasl.oauthbearer.assertion.claim.issuer,
+                    rk->rk_conf.sasl.oauthbearer.assertion.claim.audience,
+                    rk->rk_conf.sasl.oauthbearer.assertion.claim.not_before_s,
+                    rk->rk_conf.sasl.oauthbearer.assertion.claim.expiration_s,
+                    rk->rk_conf.sasl.oauthbearer.assertion.claim.jti_include);
+        }
+
+        if (!jwt_assertion) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk, "Failed to create JWT assertion");
+                goto done;
+        }
+
+        request_body = rd_kafka_oidc_jwt_bearer_build_request_body(
+            jwt_assertion, rk->rk_conf.sasl.oauthbearer.scope);
+
+        if (!request_body) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk, "Failed to build JWT request body");
+                goto done;
+        }
+
+        headers = curl_slist_append(
+            headers, "Content-Type: application/x-www-form-urlencoded");
+        headers = curl_slist_append(headers, "Accept: application/json");
+
+        herr = rd_http_post_expect_json(
+            rk, rk->rk_conf.sasl.oauthbearer.token_endpoint_url, headers,
+            request_body, strlen(request_body), timeout_s, retry, retry_ms,
+            &json);
+
+        if (unlikely(herr != NULL)) {
+                rd_kafka_log(
+                    rk, LOG_ERR, "JWT",
+                    "Failed to retrieve JWT token from \"%s\": %s (%d)",
+                    rk->rk_conf.sasl.oauthbearer.token_endpoint_url,
+                    herr->errstr, herr->code);
+                rd_kafka_oauthbearer_set_token_failure(rk, herr->errstr);
+                rd_http_error_destroy(herr);
+                goto done;
+        }
+
+        jwt_token = rd_kafka_oidc_token_try_validate(json, "access_token", &sub,
```

Comment:
Please add a comment here why we try validating with access_token first and then id_token afterwards if it fails.